backend/src/main/java/ai/verta/modeldb/lineage/LineageServiceImpl.java

```diff
@@ -187,56 +206,129 @@ public RepositoryEntity getKey() {
-          }
-          blobResult = result.get(commitSha);
-          String stringFromProtoObject =
-              ModelDBUtils.getStringFromProtoObject(
-                  Location.newBuilder().addAllLocation(blob.getLocationList()));
-          if (!blobResult.contains(stringFromProtoObject)) {
-            blobDAO.getCommitComponent(session1 -> repo, commitSha, blob.getLocationList());
-            blobResult.add(stringFromProtoObject);
-          }
-          break;
-        default:
-          throw new ModelDBException("Unexpected type", Code.INTERNAL);
-      }
+      validate(session, experimentRuns, blobs, lineageEntry);
+    }
+  }
+
+  private void validate(Session session, Set<String> experimentRuns,
+      Map<Long, RepositoryContainer> blobs, LineageEntry lineageEntry)
+      throws InvalidProtocolBufferException, ModelDBException, NoSuchAlgorithmException {
+    switch (lineageEntry.getDescriptionCase()) {
+      case EXPERIMENT_RUN:
+        String experimentRun = lineageEntry.getExperimentRun();
+        if (!experimentRuns.contains(experimentRun)) {
+          ExperimentRunEntity experimentRunEntity =
+              experimentDAO.getExperimentRun(session, experimentRun);
+          experimentRuns.add(experimentRun);
+          // Validate if current user has access to the entity or not
+          roleService.validateEntityUserWithUserInfo(
+              ModelDBServiceResourceTypes.PROJECT,
+              experimentRunEntity.getProject_id(),
+              ModelDBServiceActions.READ);
+        }
+        break;
+      case BLOB:
+        VersioningLineageEntry blob = lineageEntry.getBlob();
+        long repositoryId = blob.getRepositoryId();
+        RepositoryEntity repo;
+        CommitMap result;
+        if (!blobs.containsKey(repositoryId)) {
+          repo =
+              repositoryDAO.getRepositoryById(
+                  session, RepositoryIdentification.newBuilder().setRepoId(repositoryId).build());
+          blobs.put(repositoryId, new RepositoryContainer(repo));
+          result = blobs.get(repositoryId).getValue();
+        } else {
+          RepositoryContainer entityMapEntry = blobs.get(repositoryId);
+          repo = entityMapEntry.getKey();
+          result = entityMapEntry.getValue();
+        }
+        String commitSha = blob.getCommitSha();
+        Set<String> blobResult;
+        if (!result.containsKey(commitSha)) {
+          commitDAO.getCommitEntity(session, commitSha, session2 -> repo);
+          result.put(commitSha, new HashSet<>());
+        }
+        blobResult = result.get(commitSha);
+        String stringFromProtoObject =
+            ModelDBUtils.getStringFromProtoObject(
+                Location.newBuilder().addAllLocation(blob.getLocationList()));
+        if (!blobResult.contains(stringFromProtoObject)) {
+          blobDAO.getCommitComponent(session2 -> repo, commitSha, blob.getLocationList());
+          blobResult.add(stringFromProtoObject);
+        }
```

Comment:
It's not clear to me what this is doing. The previous case was just checking auth, which was straightforward. This seems to be building some sort of cache for things.

1. Why are we building a cache for more than repositories?
2. Where is the call to verify permission to the repository?

Reply:
1. It's a cache that exists only during a request. There can be multiple entries with same commit hash, but different location so that's why cache here can be useful. For locations cache is probably not needed, so I have removed it now.
2. getRepositoryById checks permissions. Added a comment